### PR TITLE
Add check for disabled CSRF protection in Spring

### PR DESCRIPTION
**Pull request by ggolawski**
Spring has built-in CSRF protection. However, it's possible to disable it, which most likely makes the application vulnerable to Cross-Site Request Forgery (CWE-352).

Spring CSRF protection can be disabled via Java configuration (more details [here](https://docs.spring.io/spring-security/site/docs/current/reference/html5/#servlet-csrf-configure-disable)):

```java
@Configuration
@EnableWebSecurity
public class WebSecurityConfig extends WebSecurityConfigurerAdapter {
  @Override
  protected void configure(HttpSecurity http) {
    http
      .csrf(csrf ->
        csrf.disable()
      );
  }
}
```

This PR adds a CodeQL check which finds all invocations of `CsrfConfigurer.disable()` method.

### CODE REVIEW

File: java/ql/src/Security/CWE/CWE-352/SpringCSRFProtection.qhelp [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+result in exposure of data or unintended code execution.</p>
+</overview>
+
+<recommendation>
+<p>Cross-Site Request Forgery (CSRF) protection is enabled by default in Spring with Java 
+configuration. It's recommended to not disable this.</p>
+</recommendation>
+
+<example>
+<p>The following example shows the Spring Java configuration with CSRF protection disabled.</p>
```
Comment (felicitymay) [inline java/ql/src/Security/CWE/CWE-352/SpringCSRFProtection.qhelp]: Possibly worth either mentioning how to correct the bad Java configuration, or adding something like this: "This type of configuration should only be used if you are creating a service that is used only by non-browser clients."

File: java/ql/src/Security/CWE/CWE-352/SpringCSRFProtection.qhelp [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+<sample src="SpringCSRFProtection.java" />
+</example>
+
+<references>
+<li>
+CWE:
+<a href="https://cwe.mitre.org/data/definitions/352.html">CWE-352: Cross-Site Request Forgery (CSRF)</a>.
+</li>
+<li>
```
Comment (felicitymay) [inline java/ql/src/Security/CWE/CWE-352/SpringCSRFProtection.qhelp]: This is a good reference, but should be omitted. When the help is built, a link to each CWE definition that is listed in the query metadata is added automatically.

File: java/ql/src/Security/CWE/CWE-352/SpringCSRFProtection.qhelp [hunk 3]
```diff
@@ -0,0 +1,39 @@
+<!DOCTYPE qhelp SYSTEM "qhelp.dtd">
+<qhelp>
+
+<overview>
+<p>When a web server is designed to receive a request from a client without any mechanism 
+for verifying that it was intentionally sent, then it might be possible for an attacker 
+to trick a client into making an unintentional request to the web server which will be treated 
+as an authentic request. This can be done via a URL, image load, XMLHttpRequest, etc. and can
```
[the remark this answers is not in the document]
Comment (ggolawski) [reply java/ql/src/Security/CWE/CWE-352/SpringCSRFProtection.qhelp]: I agree. I'll split it up as you suggested.

File: java/ql/src/Security/CWE/CWE-352/SpringCSRFProtection.qhelp [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+<p>When a web server is designed to receive a request from a client without any mechanism 
+for verifying that it was intentionally sent, then it might be possible for an attacker 
+to trick a client into making an unintentional request to the web server which will be treated 
+as an authentic request. This can be done via a URL, image load, XMLHttpRequest, etc. and can 
+result in exposure of data or unintended code execution.</p>
+</overview>
+
+<recommendation>
+<p>Cross-Site Request Forgery (CSRF) protection is enabled by default in Spring with Java 
+configuration. It's recommended to not disable this.</p>
```
[the remark this answers is not in the document]
Comment (ggolawski) [reply java/ql/src/Security/CWE/CWE-352/SpringCSRFProtection.qhelp]: I agree. I'll change this sentence.

File: java/ql/src/Security/CWE/CWE-352/SpringCSRFProtection.qhelp (hunk 1 above)
Comment (ggolawski) [reply java/ql/src/Security/CWE/CWE-352/SpringCSRFProtection.qhelp]: To correct this, one should not invoke csrf.disable() at all. I'll add: "This type of configuration should only be used if you are creating a service that is used only by non-browser clients."

File: java/ql/src/Security/CWE/CWE-352/SpringCSRFProtection.ql [hunk 5]
```diff
@@ -0,0 +1,22 @@
+/**
+ * @name Disabled Spring CSRF protection
+ * @description Disabling CSRF protection makes the application vulnerable to
+ *              Cross-Site Request Forgery (CSRF) attack.
```
[the remark this answers is not in the document]
Comment (ggolawski) [reply java/ql/src/Security/CWE/CWE-352/SpringCSRFProtection.ql]: I agree. I'll apply this suggestion.

File: java/ql/src/Security/CWE/CWE-352/SpringCSRFProtection.qhelp (hunk 2 above)
Comment (ggolawski) [reply java/ql/src/Security/CWE/CWE-352/SpringCSRFProtection.qhelp]: I didn't know that it is added automatically. I'll remove it.

**Comment by ggolawski** [comment]
Thanks for reviewing this PR. I applied your suggestions.

File: java/ql/src/Security/CWE/CWE-352/SpringCSRFProtection.qhelp [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+<overview>
+<p>When you set up a web server to receive a request from a client without any mechanism
+for verifying that it was intentionally sent, then it is vulnerable to attack. An attacker can
+trick a client into making an unintended request to the web server that will be treated as
+an authentic request. This can be done via a URL, image load, XMLHttpRequest, etc. and can 
+result in exposure of data or unintended code execution.</p>
+</overview>
+
+<recommendation>
+<p>Cross-Site Request Forgery (CSRF) protection is enabled by default. Spring's recommendation
```
Comment (felicitymay) [inline java/ql/src/Security/CWE/CWE-352/SpringCSRFProtection.qhelp]: Many apologies for a poor review comment. On re-reading this, I can see that we need to mention Spring in the first sentence (which your original text did).

```suggestion
<p>When you use Spring, Cross-Site Request Forgery (CSRF) protection is enabled by default. Spring's recommendation
```

**Comment by aschackmull** [comment]
Could you also run autoformat on the QL code? (That's available in the right-click menu in VSCode)

File: java/ql/src/Security/CWE/CWE-352/SpringCSRFProtection.qhelp (hunk 6 above)
Comment (ggolawski) [reply java/ql/src/Security/CWE/CWE-352/SpringCSRFProtection.qhelp]: I also missed it. I'll fix it.

**Comment by ggolawski** [comment]
I corrected the help and formatted the QL code.

**Pull request merged**